File: DSA/done/1462 - checkIfPrerequisite.py
# https://leetcode.com/problems/course-schedule-iv/submissions/1297670987/

from typing import List
from collections import defaultdict

class Solution:
    def checkIfPrerequisite(self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:
        graph = defaultdict(list)
        for a, b in prerequisites:
            graph[b].append(a)

        visited = {}
        def dfs(node):
            if node in visited:
                return visited[node]

            prereq = set()
            for n in graph[node]:
                prereq.add(n)
                prereq.update(dfs(n))

            visited[node] = prereq
            return prereq

        for c in range(numCourses):
            if c not in visited:
                dfs(c)

        res = []
        for s, d in queries:
            res.append(s in visited[d])

        return res
                

File: DSA/done/1462 - checkIfPrerequisite.java
// https://leetcode.com/problems/course-schedule-iv/submissions/1297670987/

import java.util.Set;
import java.util.Map;
import java.util.List;
import java.util.HashMap;
import java.util.HashSet;
import java.util.ArrayList;

class Solution {
    private Map<Integer, Set<Integer>> visited = new HashMap<>();
    private Map<Integer, List<Integer>> graph = new HashMap<>();

    public List<Boolean> checkIfPrerequisite(int numCourses, int[][] prerequisites, int[][] queries) {
        for (int i = 0; i < numCourses; i++) {
            graph.put(i, new ArrayList<>());
        }
        for (int[] prerequisite : prerequisites) {
            int a = prerequisite[0];
            int b = prerequisite[1];
            graph.get(b).add(a); 
        }

        for (int c = 0; c < numCourses; c++) {
            if (!visited.containsKey(c)) {
                dfs(c);
            }
        }

        List<Boolean> result = new ArrayList<>();
        for (int[] query : queries) {
            int s = query[0];
            int d = query[1];
            result.add(visited.get(d).contains(s));
        }

        return result;
    }

    private Set<Integer> dfs(int node) {
        if (visited.containsKey(node)) {
            return visited.get(node);
        }

        Set<Integer> prereq = new HashSet<>();
        for (int neighbor : graph.get(node)) {
            prereq.add(neighbor);
            prereq.addAll(dfs(neighbor));
        }

        visited.put(node, prereq);
        return prereq;
    }
}
